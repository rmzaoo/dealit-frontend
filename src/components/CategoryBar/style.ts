import styled from "styled-components";

const Container = styled.div`
  background-color: #131a22;
  color: white;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 6px 15px;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.25);
  font-size: 13px;
  height: 25px;
  position: fixed;
  z-index: 9;
  width: 100%;
  top: 70px;
  border-bottom: 1px solid #168;

  > * {
    margin: 0 10px;
  }

  @media screen and (max-width: 768px) {
    display: none;
  }
`;

export { Container };
