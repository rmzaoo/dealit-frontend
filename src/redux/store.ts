import { legacy_createStore as createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import { fetchCategories } from "../api/categoryFetch";

const initialState = {
  user: {
    id: null,
    username: null,
    email: null,
    phone: null,
    photo: null,
    token: null,
  },
  cart: [],
  orders: [],
  addresses: [],
  creditCards: [],
  cartData: [],
  cartIsOpened: false,
  cartCombinedPrice: 0,
  categories: await fetchCategories().then((data) =>
    data.map((e: any) => {
      return {
        mainCategory: e.name,
        subCategories: e.subcategories.map((s: any) => s.name),
      };
    })
  ),
};

function reducer(
  state = initialState,
  action: { type: string; payload: any }
): any {
  switch (action.type) {
    case "SET_USER":
      return { ...state, user: action.payload };
    case "ADD":
      let check = false;
      state.cart.forEach((element: any, i: number) => {
        if (element.product.id === action.payload.product.id) {
          check = true;
        }
      });
      return !check
        ? {
            ...state,
            cart: [
              ...state.cart,
              {
                product: action.payload.product,
                quantity: action.payload.quantity,
              },
            ],
          }
        : {
            ...state,
            cart: state.cart.map((e: any) => {
              if (e.product.id === action.payload.product.id) {
                let numberQuantity = parseInt(e.quantity);
                e.quantity = numberQuantity + 1;
              }
              return e;
            }),
          };
    case "SEND_OPENED":
      return {
        ...state,
        cartIsOpened: action.payload.opened,
      };
    case "REMOVE":
      return {
        ...state,
        cart: state.cart.map((e: any) => {
          if (e.product.id === action.payload.id) {
            let numberQuantity = parseInt(e.quantity);
            e.quantity = numberQuantity - 1;
          }
          return e;
        }),
      };
    case "REMOVE_ALL":
      return {
        ...state,
        cart: state.cart.filter((e: any) => e.product.id !== action.payload.id),
      };
    case "SET_CART":
      return {
        ...state,
        cart: action.payload.cartRecovery,
      };
    case "SEND_COMBINED_PRICE":
      return {
        ...state,
        cartCombinedPrice: parseFloat(action.payload.price),
      };
    case "SET_COMBINED_PRICE":
      let sum = 0;
      state.cart.forEach((e: any) => {
        sum = sum + e.product.price * e.quantity;
      });
      return {
        ...state,
        cartCombinedPrice: sum,
      };
    case "SEND_COMBINED_PRICE_PLUS":
      return {
        ...state,
        cartCombinedPrice:
          state.cartCombinedPrice + parseFloat(action.payload.price),
      };
    case "SEND_COMBINED_PRICE_MINUS":
      return {
        ...state,
        cartCombinedPrice:
          state.cartCombinedPrice - parseFloat(action.payload.price),
      };
    default:
      return state;
  }
}

const store = createStore(reducer, composeWithDevTools(applyMiddleware(thunk)));

export default store;
