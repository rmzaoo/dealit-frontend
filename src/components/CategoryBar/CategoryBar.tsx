import React, { useLayoutEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { fetchCategories } from "../../api/productsFetch";
import TextLink from "../TextLink/TextLink";
import { Container } from "./style";

const CategoryBar = () => {
  const navigate = useNavigate();
  const categories = fetchCategories();
  
  return categories && (
    <Container>
      {categories.map((category: string) => (
        <TextLink onClick={() => navigate(`/products?category=${category}`)} key={category}>
        {category}
        </TextLink>
      ))}
    </Container>
  );
};

export default CategoryBar;
