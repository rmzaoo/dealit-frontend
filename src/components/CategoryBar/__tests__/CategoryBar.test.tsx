import { describe, test, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import CategoryBar from "../CategoryBar";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import store from "../../../redux/store";

describe("CategoryBar", () => {
  test("renders categories", () => {
    const categories = [
      "Clothing",
      "Automotive",
      "Kitchen",
      "Outdoor",
      "Gaming",
      "Electronics",
    ];


    const wrapper = ({ children }: any) => (
      <Provider store={store}>{children}</Provider>
    );

    const result = render(
      <BrowserRouter>
        <CategoryBar />
      </BrowserRouter>,
      { wrapper }
    );


    categories.forEach((category) => {
      expect(result.getByText(category));
    });

    
  });
});
