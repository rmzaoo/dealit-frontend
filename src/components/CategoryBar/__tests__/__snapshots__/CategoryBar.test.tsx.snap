// Vitest Snapshot v1

exports[`CategoryBar > renders 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gsnTZi jBqAIi"
      >
        <div
          class="sc-bczRLJ jGjlZq"
        >
          Eletronics
        </div>
        <div
          class="sc-bczRLJ jGjlZq"
        >
          Clothing
        </div>
        <div
          class="sc-bczRLJ jGjlZq"
        >
          Automotive
        </div>
        <div
          class="sc-bczRLJ jGjlZq"
        >
          Kitchen
        </div>
        <div
          class="sc-bczRLJ jGjlZq"
        >
          Outdoor
        </div>
        <div
          class="sc-bczRLJ jGjlZq"
        >
          Gaming
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gsnTZi jBqAIi"
    >
      <div
        class="sc-bczRLJ jGjlZq"
      >
        Eletronics
      </div>
      <div
        class="sc-bczRLJ jGjlZq"
      >
        Clothing
      </div>
      <div
        class="sc-bczRLJ jGjlZq"
      >
        Automotive
      </div>
      <div
        class="sc-bczRLJ jGjlZq"
      >
        Kitchen
      </div>
      <div
        class="sc-bczRLJ jGjlZq"
      >
        Outdoor
      </div>
      <div
        class="sc-bczRLJ jGjlZq"
      >
        Gaming
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CategoryBar > renders categories 1`] = `
<div
  class="sc-bczRLJ jGjlZq"
>
  category1
</div>
`;

exports[`CategoryBar > renders categories 2`] = `
<div
  class="sc-bczRLJ jGjlZq"
>
  category2
</div>
`;

exports[`CategoryBar > renders categories 3`] = `
<div
  class="sc-bczRLJ jGjlZq"
>
  category3
</div>
`;

exports[`CategoryBar > renders categories 4`] = `
<div
  class="sc-bczRLJ jGjlZq"
>
  category4
</div>
`;

exports[`CategoryBar > renders categories 5`] = `
<div
  class="sc-bczRLJ jGjlZq"
>
  category5
</div>
`;

exports[`CategoryBar > renders categories 6`] = `
<div
  class="sc-bczRLJ jGjlZq"
>
  category6
</div>
`;
